### PR TITLE
resolve static constexpr class attribute

### PR DESCRIPTION
**Pull request by Vipul-Cariappa**

### CODE REVIEW

**Comment by vgvassilev** [comment]
> Hi @vgvassilev, please look at the new changes. I am now using `Sema::InstantiateVariableDefinition` instead of `Cpp::Declare` to initialize the `static constexpr`.
> 
> I see some failing tests, that are due to the OSX dependency thing, rerunning them should fix it. I do not have the necessary permissions to rerun them.

Can you accept the invite and retry?

File: lib/Interpreter/CppInterOp.cpp [hunk 1]
```diff
@@ -1284,12 +1284,22 @@ namespace Cpp {
...
-        llvm::logAllUnhandledErrors(VDAorErr.takeError(), llvm::errs(),
-                                    "Failed to GetVariableOffset:");
-        return 0;
+      if (VDAorErr)
+        return (intptr_t)jitTargetAddressToPointer<void*>(VDAorErr.get());
+
+      llvm::logAllUnhandledErrors(VDAorErr.takeError(), llvm::errs(),
+                                  "Failed to GetVariableOffset:");
+
+      if (VD->isConstexpr() && !VD->hasInit())
```
Comment (vgvassilev) [inline lib/Interpreter/CppInterOp.cpp:1293]: Shouldn’t here the check be if the VD is a template, implicitly instantiated with no init?

Looks like we don’t have test coverage for this case though…

**Comment by Vipul-Cariappa** [comment]
Looks like cppyy-cling backend fails.

File: lib/Interpreter/CppInterOp.cpp [hunk 2]
```diff
@@ -1284,12 +1284,22 @@ namespace Cpp {
...
+        return (intptr_t)jitTargetAddressToPointer<void*>(VDAorErr.get());
+
+      llvm::logAllUnhandledErrors(VDAorErr.takeError(), llvm::errs(),
+                                  "Failed to GetVariableOffset:");
+
+      if (!VD->hasInit())
+        getSema().InstantiateVariableDefinition(SourceLocation(), VD);
+      if (VD->hasInit() &&
+          (VD->isConstexpr() || VD->getType().isConstQualified())) {
+        if (const APValue* val = VD->evaluateValue()) {
```
[the remark this answers is not in the document]
Comment (Vipul-Cariappa) [reply lib/Interpreter/CppInterOp.cpp:1297]: I did the same with https://github.com/compiler-research/CppInterOp/commit/24c82c5fdd7454d8f1eab019010d47cf77dea79c commit.
This change happened to resolve things sooner and `getSymbolAddress` was never executed. Please look at the [code coverage results](https://app.codecov.io/gh/compiler-research/CppInterOp/commit/24c82c5fdd7454d8f1eab019010d47cf77dea79c?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=compiler-research) of the same commit.

> Sorry for the force pushes

File: lib/Interpreter/CppInterOp.cpp [hunk 3]
```diff
@@ -1248,6 +1248,18 @@ namespace Cpp {
 
       if (!address)
         address = I.getAddressOfGlobal(GD);
+      if (!address) {
+        if (!VD->hasInit())
+          getSema().InstantiateVariableDefinition(SourceLocation(), VD);
```
Comment (vgvassilev) [inline lib/Interpreter/CppInterOp.cpp:1253]:
```suggestion
        if (!VD->hasInit()) {
#ifdef USE_CLING
        cling::Interpreter::PushTransactionRAII RAII(&getInterp());
#endif // USE_CLING
          getSema().InstantiateVariableDefinition(SourceLocation(), VD);
        }
```

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Pull request merged**